pkg/orchestrator/controller/controller.go

```diff
@@ -0,0 +1,262 @@
+	"k8s.io/client-go/rest"
+)
+
+type clusterWatcher struct {
+	clusterID         string
+	stopChan          chan struct{}
+	clusterWorkQueue  workqueue.RateLimitingInterface
+	endpointWorkQueue workqueue.RateLimitingInterface
+}
+
+type controller struct {
+	sync.Mutex
+	federator         federate.Federator
+	clusterWatcherMap map[string]*clusterWatcher
+
+	// Indirection hook for unit tests to supply fake client sets
+	clientSetFunc func(kubeConfig *rest.Config) (submarinerClientset.Interface, error)
+}
+
+const clusterIDLabelKey = "submariner-io/clusterID"
+
+var (
+	submarinerNamespace string
+)
+
+func init() {
+	flag.StringVar(&submarinerNamespace, "submariner-namespace", "submariner", "The namespace in which the submariner components are deployed.")
+}
+
+func New(federator federate.Federator) *controller {
+	return &controller{
+		federator:         federator,
+		clusterWatcherMap: make(map[string]*clusterWatcher),
+
+		clientSetFunc: func(c *rest.Config) (submarinerClientset.Interface, error) {
+			return submarinerClientset.NewForConfig(c)
+		},
+	}
+}
+
+func (c *controller) Start() error {
+	if err := c.federator.WatchClusters(c); err != nil {
+		return fmt.Errorf("Could not register cluster watch: %v", err)
+	}
+
+	return nil
+}
+
+func (c *controller) stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, clusterWatcher := range c.clusterWatcherMap {
+		clusterWatcher.close()
+	}
+}
+
+func (c *controller) startNewWatcherForCluster(kubeConfig *rest.Config, clusterID string) {
+	submarinerClient, err := c.clientSetFunc(kubeConfig)
+	if err != nil {
+		klog.Errorf("error building submariner clientset for cluster %s: %v", clusterID, err)
+		return
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	var watcher *clusterWatcher
+	if watcher = c.clusterWatcherMap[clusterID]; watcher != nil {
+		klog.V(2).Infof("An entry for cluster %s already exists - stopping previous watch", clusterID)
+		watcher.close()
+	}
+
+	watcher = &clusterWatcher{
+		clusterID: clusterID,
+		stopChan:  make(chan struct{}),
+		clusterWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Cluster watcher for %s", clusterID)),
+		endpointWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Endpoint watcher for %s", clusterID)),
+	}
+	c.clusterWatcherMap[clusterID] = watcher
+
+	submarinerInformerFactory := submarinerInformers.NewSharedInformerFactoryWithOptions(submarinerClient, 0,
+		submarinerInformers.WithNamespace(submarinerNamespace))
+
+	clusterInformer := submarinerInformerFactory.Submariner().V1().Clusters().Informer()
+	clusterInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.clusterWorkQueue))
+
+	endpointInformer := submarinerInformerFactory.Submariner().V1().Endpoints().Informer()
+	endpointInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.endpointWorkQueue))
```

Comment:
(unrelated to this line but the thought came here...)

There is a lot of quality code, good patterns and complex logic that is common to every watch that we need to set here. Something that would be repeated for most people doing the same (as here it is repeated for clusters and endpoints). 

I wonder if it would make sense to extrapolate these needs an provide a simpler interface for this?

Reply:
Let's open a JIRA to track it.